### DISCUSSION TITLE
Perhaps we should switch to the "Affero GPLv3 or later" license for our code

### DISCUSSION THREAD

**Discussion by specious** [littermap/littermap-aws-backend#20] [Technical]
Our [frontend](https://github.com/littermap/littermap) and [backend](https://github.com/littermap/littermap-aws-backend) code is currently [released](https://github.com/littermap/littermap-aws-backend/commit/37777a715648e1af0b914e5b9b16b1e0baca1519) as [free software](https://www.gnu.org/philosophy/free-sw.html) under the [GPL-3.0](https://www.gnu.org/licenses/quick-guide-gplv3.html) license, which is considered a strong [copyleft](https://en.wikipedia.org/wiki/Copyleft) license (a concept that originated as a [response](https://snyk.io/learn/what-is-copyleft-license/) to the increasing practice of using [copyright](https://en.wikipedia.org/wiki/Copyright) law to secure exclusive rights over proprietary software) that legally compels creators of derivative works to also in turn license the derivative work under the same license and to make the derivative source code available.

However, this provision only [comes into effect](https://www.gnu.org/licenses/quick-guide-gplv3.html#more-ways-for-developers-to-provide-source) when a version of the derivatative work is distributed outside of a private setting:

> "One of the fundamental requirements of the GPL is that when you **distribute** object code to users, you must also provide them with a way to get the source."

This positively implies that private use of derivative works [does not require](https://www.gnu.org/licenses/gpl-faq.html#GPLRequireSourcePostedPublic) the release of anything (even if copies are [distributed](https://www.gnu.org/licenses/gpl-faq.html#InternalDistribution) within the organization). In addition, original authors are not restricted from also releasing [their own work](https://www.gnu.org/licenses/gpl-faq.html#HeardOtherLicense) under other terms or licenses.

The very similar [AGPL-3](https://www.gnu.org/licenses/agpl-3.0.html) license is a slightly stronger [variant](https://fossa.com/blog/open-source-software-licenses-101-agpl-license/#the-agpl-license-an-overview) of the GPL-3 (see the [word diff](https://gplv3.fsf.org/gpl3-agpl3dd2.wdiff) between the GPL-3 and the second and final draft of the AGPL-3) that broadens the scope of the license (namely [section 13](https://www.gnu.org/licenses/agpl-3.0.html#section13) and the [preamble](https://www.gnu.org/licenses/agpl-3.0.html#preamble)) to include the requirement that source code also be made available when the derivative work is used to provide a remote service over a public network.

As explained in [this detailed overview](https://snyk.io/learn/agpl-license/):

> "The AGPL V3 license is a strong copyleft license that enforces open source on all components derived from any previous work. It closes the server-side loophole, where the source code isn’t made available if the software isn’t released. AGPL defines a user as anyone who accesses the server-side application if it is public-facing. For applications that reside inside the organization’s network, AGPL does not trigger the release of the source code."

I would like to propose to re-license the code in both of the repositories under the AGPL, because it provides stronger protection. Even code that is currently intended for client-side use would be protected from being used across a network in the future.

An arguable downside of using a license as strong as the AGPL (or GPL) is that it [forces](https://www.quora.com/Is-AGPL-a-stronger-copyleft-version-of-GPL/answer/Gil-Yehuda) any work that incorporates a derivative of the AGPL code to be released in its entirety under the stronger license, meaning that projects licensed under less restrictive licenses such as the [Apache](https://snyk.io/learn/apache-license/) license will be unlikely to incorporate code from this project, as that project would then need to become an AGPL project.

On the other hand, a strong license like the AGPL serves as powerful protection, as in the case of Truth Social having [been found to use](https://news.yahoo.com/trump-social-network-given-30-204900628.html) code from the [Mastodon](https://github.com/mastodon/mastodon) project, which is published under the AGPL, without having fully complied with its license. Publishing our code under the AGPL will likely mean that the [Software Freedom Conservancy](https://sfconservancy.org) will have an interest in helping to protect it and will [attract](https://www.researchgate.net/publication/256505756_The_Attraction_of_Contributors_in_Free_and_Open_Source_Software) community contribution. Commercial companies, however, are likely to [steer clear](https://opensource.google/documentation/reference/using/agpl-policy) of AGPL code.

A project licensed under the AGPL is more likely to be able to include code licensed under more permissive licenses, than the other way. However, it's [complicated](https://www.gnu.org/licenses/license-compatibility.html). In particular (possibly surprisingly), the GPL-3 and AGPL-3 are only [partially](https://www.gnu.org/licenses/license-list.html#AGPLv3.0) compatible with each other:

> "Please note that the GNU AGPL is not compatible with GPLv2. It is also technically not compatible with GPLv3 in a strict sense: you cannot take code released under the GNU AGPL and convey or modify it however you like under the terms of GPLv3, or vice versa. However, you are allowed to combine separate modules or source files released under both of those licenses in a single project, ..."

This convenient [chart](https://dwheeler.com/essays/floss-license-slide.html) visualizes some of the free licenses that would be subsumed by the AGPL-3 (meaning code under those licences could be included in this project and become AGPL code). Notably, as shown, the [Mozilla Public License](https://en.wikipedia.org/wiki/Mozilla_Public_License) 1.1 is incompatible with GPL code, while MPL 2.0 is known to be semi-compatible with the AGPL, as [explained here](https://en.wikipedia.org/wiki/Mozilla_Public_License#Terms):

> "The one exception to covered source files remaining under the MPL occurs when code under version 2.0 or later is combined with separate code files under the GNU GPL, GNU Lesser GPL (LGPL), or Affero GPL (AGPL). In this case, the program as a whole will be under the chosen GNU license, but the MPL-covered files will be dual-licensed, so that recipients can choose to distribute them under that GNU License or the MPL. The initial author of MPL code may choose to opt out of this GPL compatibility by adding a notice to its source files."

This [article](https://en.wikipedia.org/wiki/License_compatibility#Compatibility_of_FOSS_licenses) on Wikipedia goes into detail on license compatibility, while this in-depth [article](https://www.oreilly.com/library/view/understanding-open-source/0596005814/ch03.html) from O'Reilly explains the differences between the GPL, LGPL and the MPL (however, note that it does not refer to the latest versions of these licenses).

In light of future compatibility [nuances](https://www.gnu.org/licenses/identify-licenses-clearly.html), whereby code under this the GPL or AGPL isn't automatically future-compatible, I propose to license the code for this project as **GNU AGPL-3.0-or-later**.

It's important to note that publishing code under a particular free software license implies taking a stance on the importance of free software. A weak license ensures ample compatibility, but offers little protection. A strong license leads to sometimes unfortunate compatibility issues.

Essentially, licensing this software under the AGPL makes it into an AGPL "seed". Any software that grows from it is also going to be under the AGPL. It will stand in contrast to software that can "swing the other way".

Some projects that have chosen to publish their source code under the AGPL license:

- [Grafana](https://grafana.com/blog/2021/04/20/qa-with-our-ceo-on-relicensing/) (multiple repositories)

  See discussion: grafana/grafana/issues/34091

- [Mastodon](https://github.com/mastodon/mastodon#)

  See discussion: mastodon/mastodon/issues/49

- [Internet Archive](https://github.com/internetarchive/openlibrary/blob/master/Readme.md#license) (multiple repositories)

See [all](https://github.com/search?q=license:agpl-3.0&s=stars) AGPL-3.0 repositories on github.

**Comment by danieltoben** [comment]
I agree with this switch, and I appreciate your research into licensing!

**Comment by specious** [comment]
All of the code is published under the AGPL 3.0 or later as of these commits:

- 100f0c25197a19affd04851d9694201930bcf994
- https://github.com/littermap/littermap/commit/54ce2b45c110ce2f43bd29742ed67cdabd26d485
